sidecar-ffi/src/lib.rs

```diff
@@ -999,6 +1001,117 @@ pub unsafe extern "C" fn ddog_get_agent_info_env<'a>(
         .unwrap_or(ffi::CharSlice::empty())
 }
 
+#[macro_export]
+macro_rules! check {
+    ($failable:expr, $msg:expr) => {
+        match $failable {
+            Ok(o) => o,
+            Err(e) => {
+                tracing::error!("{}: {}", $msg, e);
+                return;
+            }
+        }
+    };
+}
+
+#[repr(C)]
+#[derive()]
+pub struct SenderParameters {
+    pub tracer_headers_tags: TracerHeaderTags<'static>,
+    pub transport: Box<SidecarTransport>,
+    pub instance_id: Box<InstanceId>,
+    pub limit: usize,
+    pub n_requests: i64,
+    pub buffer_size: i64,
+    pub url: CharSlice<'static>,
+}
+
+#[no_mangle]
+#[allow(clippy::missing_safety_doc)]
+pub unsafe extern "C" fn ddog_send_traces_to_sidecar(
+    traces: &mut TracesBytes,
+    parameters: &mut SenderParameters,
+) {
+    let size: usize = traces.iter().map(|trace| trace.len()).sum();
+
+    // Check connection to the sidecar
+    if parameters.transport.is_closed() {
+        tracing::info!(
+            "Skipping flushing traces of size {} as connection to sidecar failed",
+            size
+        );
+        return;
+    }
+
+    // Create and map shared memory
+    let shm = check!(
+        ShmHandle::new(parameters.limit),
+        "Failed to create shared memory"
+    );
+
+    let mut mapped_shm = check!(shm.clone().map(), "Failed to map shared memory");
+
+    // Write traces to the shared memory
+    let mut cursor = Cursor::new(mapped_shm.as_slice_mut());
+    let written = match rmp_serde::encode::write_named(&mut cursor, &traces) {
+        Ok(()) => cursor.position() as usize,
+        Err(_) => {
+            tracing::error!("Failed serializing the traces");
+            0
+        }
+    };
+
+    // Send traces to the sidecar via the shared memory handler
+    let mut size_hint = written;
+    if parameters.n_requests > 0 {
+        size_hint = size_hint.max((parameters.buffer_size / parameters.n_requests + 1) as usize);
+    }
+
+    let send_error = blocking::send_trace_v04_shm(
+        &mut parameters.transport,
+        &parameters.instance_id,
+        shm,
+        size_hint,
+        check!(
+            (&parameters.tracer_headers_tags).try_into(),
+            "Failed to convert tracer headers tags"
+        ),
+    );
+
+    // Retry sending traces via bytes if there was an error
+    if send_error.is_err() {
+        let mut buffer = vec![0u8; written];
+        let pointer = buffer.as_mut_slice();
+
+        let mut cursor = Cursor::new(pointer);
+        rmp_serde::encode::write_named(&mut cursor, &traces).unwrap_or_else(|_| {
+            tracing::error!("Failed serializing the traces");
+        });
+
+        match blocking::send_trace_v04_bytes(
+            &mut parameters.transport,
+            &parameters.instance_id,
+            buffer.as_mut_slice().to_vec(),
```

Comment:
```suggestion
            buffer,
```
Otherwise you duplicate the vec